src/services/db/TokenService.ts

```diff
@@ -0,0 +1,340 @@
+
+import { AccessToken } from "@database/models"
+
+// Env vars
+export const GITHUB_TOKEN_LIMIT = 5000
+export const GITHUB_TOKEN_THRESHOLD = 4000 // allowed uses
+export const GITHUB_RESET_INTERVAL = 60 * 60 // seconds
+export const ACTIVE_TOKEN_ALERT_1 = 0.6
+export const ACTIVE_TOKEN_ALERT_2 = 0.8
+export const GITHUB_TOKEN_REMAINING_HEADER = "x-ratelimit-remaining"
+export const GITHUB_TOKEN_RESET_HEADER = "x-ratelimit-reset"
+
+export type MaybeResetTime = Result<number, null>
+export const NoResetTime: MaybeResetTime = err(null)
+
+export type TokenData = {
+  id: number
+
+  tokenString: string
+
+  remainingRequests: number
+
+  resetTime: MaybeResetTime
+}
+
+export type MaybeTokenData = Result<TokenData, null>
+export const NoTokenData: MaybeTokenData = err(null)
+
+type ResponseTokenData = {
+  token: string
+  remainingRequests: number
+  resetTime: number
+}
+
+export function queryActiveTokens(
+  tokenDB: ModelStatic<AccessToken>
+): ResultAsync<TokenData[], DatabaseError> {
+  return fromPromise(
+    tokenDB.findAll({
+      where: {
+        isReserved: false,
+      },
+    }),
+    (error) =>
+      new DatabaseError("Unable to retrieve active tokens from database")
+  ).map((activeTokens) =>
+    activeTokens.map((activeToken) => ({
+      id: activeToken.id,
+      tokenString: activeToken.token,
+      remainingRequests: GITHUB_TOKEN_LIMIT,
+      resetTime: NoResetTime,
+    }))
+  )
+}
+
+type LoggerType = typeof logger
+
+export function activeUsageAlert(
+  activeTokensData: TokenData[],
+  useLogger: LoggerType
+) {
+  let exhaustedTokensCount = 0
+  activeTokensData.forEach((tokenData) => {
+    if (
+      tokenData.remainingRequests <=
+      GITHUB_TOKEN_LIMIT - GITHUB_TOKEN_THRESHOLD
+    ) {
+      exhaustedTokensCount += 1
+    }
+  })
+  if (exhaustedTokensCount >= activeTokensData.length * ACTIVE_TOKEN_ALERT_2) {
+    useLogger.info(`${ACTIVE_TOKEN_ALERT_2}% of access token capacity reached`)
+    useLogger.info(
+      `${exhaustedTokensCount}/${activeTokensData.length} active tokens exhausted`
+    )
+  } else if (
+    exhaustedTokensCount >=
+    activeTokensData.length * ACTIVE_TOKEN_ALERT_1
+  ) {
+    useLogger.info(`${ACTIVE_TOKEN_ALERT_1}% of access token capacity reached`)
+    useLogger.info(
+      `${exhaustedTokensCount}/${activeTokensData.length} active tokens exhausted`
+    )
+  }
+}
+
+export function parseResponseTokenData(
+  response: AxiosResponse
+): Result<ResponseTokenData, TokenParsingError> {
+  // response.config.headers.Authorization format: token ghp_********************************
+  if (
+    typeof response.config?.headers?.Authorization !== "string" ||
+    response.config?.headers?.Authorization.length !== 46 ||
+    response.config?.headers?.Authorization.slice(0, 6) !== "token " ||
+    Number.isNaN(+response.headers?.[GITHUB_TOKEN_REMAINING_HEADER]) ||
+    Number.isNaN(+response.headers?.[GITHUB_TOKEN_RESET_HEADER])
```

Comment:
style: shall we shift this to a function or something like 

isGithubToken = ...
if (!isGithubToken) {
    return err(...)
}

Reply:
Make sense! Will do so.